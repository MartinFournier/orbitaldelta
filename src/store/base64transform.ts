import { createTransform } from 'redux-persist';
import { TransformConfig } from 'redux-persist/lib/createTransform';

function transform(config?: TransformConfig) {
  return createTransform(
    (inboundState: string) => window.btoa(encodeURIComponent(inboundState)),
    (outboundState: string) => decodeURIComponent(window.atob(outboundState)),
    config,
  );
}

export default transform;
