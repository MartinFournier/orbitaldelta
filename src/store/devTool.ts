import { playerActions } from 'app/player/slice';

/**
 * These entries will not appear in the redux dev tool. Used to prevent spammy actions from flooding the dev tool.
 */
export const devtoolBlackListedActions = [
  playerActions.incrementProcessingDeltaMs.type,
  playerActions.changeLastProcessed.type,
];
