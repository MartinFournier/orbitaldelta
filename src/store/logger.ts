import { playerActions } from 'app/player/slice';
import { createLogger } from 'redux-logger';

const ignoredActions = [
  playerActions.incrementProcessingDeltaMs.type,
  playerActions.changeLastProcessed.type,
];

export default createLogger({
  predicate: (getState, action) => !ignoredActions.includes(action.type),
});
