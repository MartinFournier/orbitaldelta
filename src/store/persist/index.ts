import partial from 'lodash/partial';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';

import { createBlacklistFilter } from 'redux-persist-transform-filter';

import storage from 'redux-persist/lib/storage';
import createCompressor from 'redux-persist-transform-compress';
import { envIsDev } from 'utilities/environment';

const blacklistFilter = createBlacklistFilter('engine', ['isGameSaving']);
const lzCompressor = createCompressor();
const transforms = [blacklistFilter];
if (!envIsDev) transforms.push(lzCompressor);

export const persistConfig = {
  key: 'root',
  keyPrefix: 'orbital',
  version: 1,
  storage,
  transforms,
  throttle: 100,
  debug: envIsDev,
};

export const getPersistReducer = partial(persistReducer, persistConfig);
export const getPersistStore = persistStore;

export const ignoredSerializableActions = [
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
];
