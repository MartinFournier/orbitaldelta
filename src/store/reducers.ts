import { combineReducers } from '@reduxjs/toolkit';
import playerReducer from 'app/components/Player/slice';

const rootReducer = combineReducers({
  player: playerReducer,
});

export default rootReducer;
