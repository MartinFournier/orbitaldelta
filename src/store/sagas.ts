import { all, call, spawn } from 'redux-saga/effects';

import { autosave, watchEngineRehydrate } from 'app/engine/slice/saga';
import { watchGameTimeProcess } from 'app/game-time/slice/saga';
import { watchPlayerProcess } from 'app/player/slice/saga';

export default function* rootSaga() {
  const sagas = [
    autosave,
    watchGameTimeProcess,
    watchEngineRehydrate,
    watchPlayerProcess,
  ];

  yield all(
    sagas.map(saga =>
      spawn(function* () {
        while (true) {
          try {
            yield call(saga);
            break;
          } catch (e) {
            console.error(e);
          }
        }
      }),
    ),
  );
}
