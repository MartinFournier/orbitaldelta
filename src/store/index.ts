import { configureStore } from '@reduxjs/toolkit';

import createSagaMiddleware from 'redux-saga';

import rootReducer, { RootReducerType } from './reducers';
import rootSaga from './sagas';
import { devtoolBlackListedActions } from './devTool';
import {
  getPersistReducer,
  getPersistStore,
  ignoredSerializableActions,
} from './persist';

function setupStore() {
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  const store = configureStore({
    reducer: getPersistReducer(rootReducer),
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: ignoredSerializableActions,
        },
      }).concat(sagaMiddleware),
    devTools: { actionsBlacklist: devtoolBlackListedActions },
  });

  sagaMiddleware.run(rootSaga);

  const persistor = getPersistStore(store);
  return { store, persistor };
}

const { store, persistor } = setupStore();
export { store, persistor };

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;
