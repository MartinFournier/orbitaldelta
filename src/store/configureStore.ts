import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage';
import { persistReducer, persistStore } from 'redux-persist';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './reducers';
import rootSaga from './sagas';
import logger from './logger';

const persistConfig = {
  key: 'player',
  storage,
};

function configureAppStore() {
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const middleware = (getDefaultMiddleware: any) => {
    const m = getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: ['persist/PERSIST', 'persist/HYDRATE', 'persist/FLUSH'],
      },
    }).concat(...[sagaMiddleware]);
    if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test') {
      m.push(logger);
    }
    return m;
  };

  const store = configureStore({
    reducer: persistReducer(persistConfig, rootReducer),
    middleware: middleware,
    devTools: process.env.NODE_ENV !== 'production',
  });

  sagaMiddleware.run(rootSaga);

  return store;
}

const store = configureAppStore();
const persistor = persistStore(store);

export { store, persistor };
