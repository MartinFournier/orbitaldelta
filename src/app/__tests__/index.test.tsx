import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { persistor } from 'store';
import { App } from '../index';

const renderer = createRenderer();

describe('<App />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(<App persistor={persistor} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
