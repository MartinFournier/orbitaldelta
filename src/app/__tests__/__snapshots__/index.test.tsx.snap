// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match the snapshot 1`] = `
<BrowserRouter>
  <Helmet
    defaultTitle="Orbital Δ"
    defer={true}
    encodeSpecialCharacters={true}
    htmlAttributes={
      Object {
        "lang": "en-US",
      }
    }
    prioritizeSeoTags={false}
    titleTemplate="[dev] %s - Orbital Δ - vundefined (undefined)"
  >
    <meta
      content="An incremental based on automated rocketry."
      name="description"
    />
  </Helmet>
  <AppThemeProvider>
    <ErrorHandler>
      <PersistGate
        loading={
          <LoadingPage
            category="dev"
            icon={<Memo />}
            route="/loading"
            title="Loading"
          />
        }
        persistor={
          Object {
            "dispatch": [Function],
            "flush": [Function],
            "getState": [Function],
            "pause": [Function],
            "persist": [Function],
            "purge": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(Symbol.observable): [Function],
          }
        }
      >
        <GlobalHotkeys>
          <StyledSnackbarProvider>
            <SnackbarUtilsConfigurator />
            <Engine>
              <AppRoutes
                persistor={
                  Object {
                    "dispatch": [Function],
                    "flush": [Function],
                    "getState": [Function],
                    "pause": [Function],
                    "persist": [Function],
                    "purge": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                    Symbol(Symbol.observable): [Function],
                  }
                }
              />
            </Engine>
          </StyledSnackbarProvider>
        </GlobalHotkeys>
      </PersistGate>
    </ErrorHandler>
  </AppThemeProvider>
</BrowserRouter>
`;
